File: deepskilling/M2_Data Structures_and_Algorithms/TaskManagementSystem.java
public class TaskManagementSystem {
    static class Task {
        String taskId;
        String taskName;
        String status;
        Task next;

        Task(String taskId, String taskName, String status) {
            this.taskId = taskId;
            this.taskName = taskName;
            this.status = status;
            this.next = null;
        }

        @Override
        public String toString() {
            return String.format("Task{id=%s, name=%s, status=%s}", taskId, taskName, status);
        }
    }

    private Task head;

    public void addTask(Task task) {
        if (head == null) {
            head = task;
            return;
        }
        Task current = head;
        while (current.next != null) {
            current = current.next;
        }
        current.next = task;
    }

    public Task searchTask(String taskId) {
        Task current = head;
        while (current != null) {
            if (current.taskId.equals(taskId)) {
                return current;
            }
            current = current.next;
        }
        return null;
    }

    public void traverseTasks() {
        if (head == null) {
            System.out.println("No tasks found.");
            return;
        }
        Task current = head;
        while (current != null) {
            System.out.println(current);
            current = current.next;
        }
    }

    public boolean deleteTask(String taskId) {
        if (head == null) {
            return false;
        }
        if (head.taskId.equals(taskId)) {
            head = head.next;
            return true;
        }
        Task current = head;
        while (current.next != null) {
            if (current.next.taskId.equals(taskId)) {
                current.next = current.next.next;
                return true;
            }
            current = current.next;
        }
        return false;
    }

    public static void main(String[] args) {
        TaskManagementSystem taskSystem = new TaskManagementSystem();

        taskSystem.addTask(new Task("T100", "Design module", "In Progress"));
        taskSystem.addTask(new Task("T101", "Write tests", "Pending"));
        taskSystem.addTask(new Task("T102", "Deploy app", "Pending"));

        System.out.println("All tasks:");
        taskSystem.traverseTasks();

        System.out.println("\nSearch for T101:");
        System.out.println(taskSystem.searchTask("T101"));

        taskSystem.deleteTask("T100");
        System.out.println("\nAfter deleting T100:");
        taskSystem.traverseTasks();
    }
}

File: deepskilling/M2_Data Structures_and_Algorithms/TaskManagementSystem.cs
using System;

public class TaskManagementSystem
{
    public class Task
    {
        public string TaskId { get; set; }
        public string TaskName { get; set; }
        public string Status { get; set; }
        public Task Next { get; set; }

        public Task(string taskId, string taskName, string status)
        {
            TaskId = taskId;
            TaskName = taskName;
            Status = status;
            Next = null;
        }

        public override string ToString()
        {
            return $"Task{{id={TaskId}, name={TaskName}, status={Status}}}";
        }
    }

    private Task _head;

    public void AddTask(Task task)
    {
        if (_head == null)
        {
            _head = task;
            return;
        }
        Task current = _head;
        while (current.Next != null)
        {
            current = current.Next;
        }
        current.Next = task;
    }

    public Task SearchTask(string taskId)
    {
        Task current = _head;
        while (current != null)
        {
            if (string.Equals(current.TaskId, taskId, StringComparison.Ordinal))
            {
                return current;
            }
            current = current.Next;
        }
        return null;
    }

    public void TraverseTasks()
    {
        if (_head == null)
        {
            Console.WriteLine("No tasks found.");
            return;
        }
        Task current = _head;
        while (current != null)
        {
            Console.WriteLine(current);
            current = current.Next;
        }
    }

    public bool DeleteTask(string taskId)
    {
        if (_head == null)
        {
            return false;
        }
        if (string.Equals(_head.TaskId, taskId, StringComparison.Ordinal))
        {
            _head = _head.Next;
            return true;
        }
        Task current = _head;
        while (current.Next != null)
        {
            if (string.Equals(current.Next.TaskId, taskId, StringComparison.Ordinal))
            {
                current.Next = current.Next.Next;
                return true;
            }
            current = current.Next;
        }
        return false;
    }

    public static void Main(string[] args)
    {
        TaskManagementSystem taskSystem = new TaskManagementSystem();
        taskSystem.AddTask(new Task("T100", "Design module", "In Progress"));
        taskSystem.AddTask(new Task("T101", "Write tests", "Pending"));
        taskSystem.AddTask(new Task("T102", "Deploy app", "Pending"));

        Console.WriteLine("All tasks:");
        taskSystem.TraverseTasks();

        Console.WriteLine("\nSearch for T101:");
        Console.WriteLine(taskSystem.SearchTask("T101"));

        taskSystem.DeleteTask("T100");

        Console.WriteLine("\nAfter deleting T100:");
        taskSystem.TraverseTasks();
    }
}
